In [10]:
import pandas as pd
import numpy as np

%matplotlib inline 
import matplotlib.pyplot as plt   


#importing seaborn for statistical plots
import seaborn as sns

# 1. Import the required libraries and read the dataset

In [11]:
bank=pd.read_csv("E:\\GLC\\python\\assignment\\Bank_Personal_Loan_Modelling project.csv")

In [12]:
bank

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  \
0        1   25           1      49     91107       4    1.6          1   
1        2   45          19      34     90089       3    1.5          1   
2        3   39          15      11     94720       1    1.0          1   
3        4   35           9     100     94112       1    2.7          2   
4        5   35           8      45     91330       4    1.0          2   
...    ...  ...         ...     ...       ...     ...    ...        ...   
4995  4996   29           3      40     92697       1    1.9          3   
4996  4997   30           4      15     92037       4    0.4          1   
4997  4998   63          39      24     93023       2    0.3          3   
4998  4999   65          40      49     90034       3    0.5          2   
4999  5000   28           4      83     92612       3    0.8          1   

      Mortgage  Personal Loan  Securities Account  CD Account  Online  \
0            0              0                   1           0       0   
1            0              0                   1           0       0   
2            0              0                   0           0       0   
3            0              0                   0           0       0   
4            0              0                   0           0       0   
...        ...            ...                 ...         ...     ...   
4995         0              0                   0           0       1   
4996        85              0                   0           0       1   
4997         0              0                   0           0       0   
4998         0              0                   0           0       1   
4999         0              0                   0           0       1   

      CreditCard  
0              0  
1              0  
2              0  
3              0  
4              1  
...          ...  
4995           0  
4996           0  
4997           0  
4998           0  
4999           1  

[5000 rows x 14 columns]

# 2. Check the first few samples, shape, info of the data and try to familiarize yourself with different features.

In [13]:
bank.head(2)

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0

In [14]:
bank.tail(2)

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  \
4998  4999   65          40      49     90034       3    0.5          2   
4999  5000   28           4      83     92612       3    0.8          1   

      Mortgage  Personal Loan  Securities Account  CD Account  Online  \
4998         0              0                   0           0       1   
4999         0              0                   0           0       1   

      CreditCard  
4998           0  
4999           1

In [15]:
bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


In [21]:
bank.shape

(5000, 14)

# 3. Check if there are any duplicate records present in the dataset? If yes, drop them. and Drop the columns which you feel are redundant

In [16]:
bank.isnull().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [20]:
bank.duplicated().sum() # another way to find duplicates in given dataset

0

# 4. Display the Five Point Summary and write your key findings.

In [18]:
bank.describe()

ID          Age   Experience       Income      ZIP Code  \
count  5000.000000  5000.000000  5000.000000  5000.000000   5000.000000   
mean   2500.500000    45.338400    20.104600    73.774200  93152.503000   
std    1443.520003    11.463166    11.467954    46.033729   2121.852197   
min       1.000000    23.000000    -3.000000     8.000000   9307.000000   
25%    1250.750000    35.000000    10.000000    39.000000  91911.000000   
50%    2500.500000    45.000000    20.000000    64.000000  93437.000000   
75%    3750.250000    55.000000    30.000000    98.000000  94608.000000   
max    5000.000000    67.000000    43.000000   224.000000  96651.000000   

            Family        CCAvg    Education     Mortgage  Personal Loan  \
count  5000.000000  5000.000000  5000.000000  5000.000000    5000.000000   
mean      2.396400     1.937938     1.881000    56.498800       0.096000   
std       1.147663     1.747659     0.839869   101.713802       0.294621   
min       1.000000     0.000000     1.000000     0.000000       0.000000   
25%       1.000000     0.700000     1.000000     0.000000       0.000000   
50%       2.000000     1.500000     2.000000     0.000000       0.000000   
75%       3.000000     2.500000     3.000000   101.000000       0.000000   
max       4.000000    10.000000     3.000000   635.000000       1.000000   

       Securities Account  CD Account       Online   CreditCard  
count         5000.000000  5000.00000  5000.000000  5000.000000  
mean             0.104400     0.06040     0.596800     0.294000  
std              0.305809     0.23825     0.490589     0.455637  
min              0.000000     0.00000     0.000000     0.000000  
25%              0.000000     0.00000     0.000000     0.000000  
50%              0.000000     0.00000     1.000000     0.000000  
75%              0.000000     0.00000     1.000000     1.000000  
max              1.000000     1.00000     1.000000     1.000000

# 5. There are negative values in the variable 'Experience'. Convert them to non-negative values. (Hint: .abs function)

In [29]:
bank['Experience'].tail(2)

4998    40
4999     4
Name: Experience, dtype: int64

In [30]:
bank['Experience'].head(2)

0     1
1    19
Name: Experience, dtype: int64

In [31]:
bank

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  \
0        1   25           1      49     91107       4    1.6          1   
1        2   45          19      34     90089       3    1.5          1   
2        3   39          15      11     94720       1    1.0          1   
3        4   35           9     100     94112       1    2.7          2   
4        5   35           8      45     91330       4    1.0          2   
...    ...  ...         ...     ...       ...     ...    ...        ...   
4995  4996   29           3      40     92697       1    1.9          3   
4996  4997   30           4      15     92037       4    0.4          1   
4997  4998   63          39      24     93023       2    0.3          3   
4998  4999   65          40      49     90034       3    0.5          2   
4999  5000   28           4      83     92612       3    0.8          1   

      Mortgage  Personal Loan  Securities Account  CD Account  Online  \
0            0              0                   1           0       0   
1            0              0                   1           0       0   
2            0              0                   0           0       0   
3            0              0                   0           0       0   
4            0              0                   0           0       0   
...        ...            ...                 ...         ...     ...   
4995         0              0                   0           0       1   
4996        85              0                   0           0       1   
4997         0              0                   0           0       0   
4998         0              0                   0           0       1   
4999         0              0                   0           0       1   

      CreditCard  
0              0  
1              0  
2              0  
3              0  
4              1  
...          ...  
4995           0  
4996           0  
4997           0  
4998           0  
4999           1  

[5000 rows x 14 columns]

In [32]:
for value in bank['Experience']:
    if value < 0:
        print("Negative value found!")

Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative value found!
Negative v

In [33]:
# Count the number of negative values in a specific column
count_neg = (bank['Experience'] < 0).sum()

print(f'The column contains {count_neg} negative values')

The column contains 52 negative values


In [34]:
count_neg

52

In [35]:
# Create a Boolean Series indicating whether each row has negative values
has_neg = bank.lt(0).any(axis=1)

# Get the indices of the rows with negative values
neg_indices = bank.index[has_neg]

print(f'The following indices have negative values: {neg_indices}')

The following indices have negative values: Int64Index([  89,  226,  315,  451,  524,  536,  540,  576,  583,  597,  649,
             670,  686,  793,  889,  909, 1173, 1428, 1522, 1905, 2102, 2430,
            2466, 2545, 2618, 2717, 2848, 2876, 2962, 2980, 3076, 3130, 3157,
            3279, 3284, 3292, 3394, 3425, 3626, 3796, 3824, 3887, 3946, 4015,
            4088, 4116, 4285, 4411, 4481, 4514, 4582, 4957],
           dtype='int64')


#### Then, it creates a Boolean Series has_neg indicating whether each row in the DataFrame has at least one negative value. The lt() function is used to compare each value in the DataFrame to zero, and the any() function is used to check whether any of the values in each row are negative. The axis=1 parameter specifies that we are checking for negative values row-wise. Finally, the code gets the indices of the rows with negative values using the .index attribute of the DataFrame and the Boolean Series has_neg. The indices are stored in the neg_indices variable, and then printed using an f-string. This will show you the indices of the rows that contain at least one negative value.

In [39]:
bank.Experience.abs()

0        1
1       19
2       15
3        9
4        8
        ..
4995     3
4996     4
4997    39
4998    40
4999     4
Name: Experience, Length: 5000, dtype: int64

In [41]:
# Make a copy of the column to compare with later
orig_col = bank['Experience'].copy()

# Apply the abs() function to the column
bank['Experience'] = bank['Experience'].abs()

# Compare the original column with the modified column
if (bank['Experience'] == orig_col.abs()).all():
    print('All negative values have been changed')
else:
    print('Some negative values have not been changed')

All negative values have been changed


In [42]:
# Create a Boolean Series indicating whether each row has negative values
has_neg = bank.lt(0).any(axis=1)

# Get the indices of the rows with negative values
neg_indices = bank.index[has_neg]

print(f'The following indices have negative values: {neg_indices}')

The following indices have negative values: Int64Index([], dtype='int64')


In [43]:
# Count the number of negative values in a specific column
count_neg = (bank['Experience'] < 0).sum()

print(f'The column contains {count_neg} negative values')

The column contains 0 negative values


# 6. Get the target column distribution and comment on the class distribution.

In [44]:
bank.head(5)

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [45]:
# Get the value counts of the target column
target_counts = bank['Personal Loan'].value_counts()

# Print the value counts
print(target_counts)

# Calculate the percentage of each class
class_percents = target_counts / target_counts.sum() * 100

# Print the class percentages
print(class_percents)

0    4520
1     480
Name: Personal Loan, dtype: int64
0    90.4
1     9.6
Name: Personal Loan, dtype: float64


'''It then gets the value counts of the "Personal Loan" column using the .value_counts() method, which counts the number of occurrences of each unique value in the column. The value counts and class percentages are then printed to the console.

   -->We can see that the "Personal Loan" column contains two classes: 0 and 1. Class 0 has 4520 instances, which is about 90.4% of the total, and class 1 has 480 instances, which is about 9.6% of the total. 
            -->The class distribution of this target variable is imbalanced, with class 0 being the majority class and class 1 being the minority class. This means that any model trained on this data may have a bias toward predicting class 0. Special care should be taken when building a predictive model using this dataset to ensure that the minority class is not overlooked.'''

# 7. Store the target column (i.e.Personal Loan) in the y variable and the rest of the columns in the Xvariable.

In [46]:

# store the target column in a variable y
y = bank['Personal Loan']

# store the rest of the columns in a variable X
X = bank.drop('Personal Loan', axis=1)


In [47]:
y

0       0
1       0
2       0
3       0
4       0
       ..
4995    0
4996    0
4997    0
4998    0
4999    0
Name: Personal Loan, Length: 5000, dtype: int64

In [48]:
X

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  \
0        1   25           1      49     91107       4    1.6          1   
1        2   45          19      34     90089       3    1.5          1   
2        3   39          15      11     94720       1    1.0          1   
3        4   35           9     100     94112       1    2.7          2   
4        5   35           8      45     91330       4    1.0          2   
...    ...  ...         ...     ...       ...     ...    ...        ...   
4995  4996   29           3      40     92697       1    1.9          3   
4996  4997   30           4      15     92037       4    0.4          1   
4997  4998   63          39      24     93023       2    0.3          3   
4998  4999   65          40      49     90034       3    0.5          2   
4999  5000   28           4      83     92612       3    0.8          1   

      Mortgage  Securities Account  CD Account  Online  CreditCard  
0            0                   1           0       0           0  
1            0                   1           0       0           0  
2            0                   0           0       0           0  
3            0                   0           0       0           0  
4            0                   0           0       0           1  
...        ...                 ...         ...     ...         ...  
4995         0                   0           0       1           0  
4996        85                   0           0       1           0  
4997         0                   0           0       0           0  
4998         0                   0           0       1           0  
4999         0                   0           0       1           1  

[5000 rows x 13 columns]

# 8. Split the dataset into two parts (i.e. 70% train and 30% test). and standardize the columns using the z-score scaling approach.

In [49]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Split the dataset into 70% train and 30% test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Standardize the columns of the train and test sets using z-score scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [50]:
X_train_scaled

array([[-0.47189175,  0.8637215 ,  0.44850537, ..., -0.2532849 ,
        -1.2189321 , -0.65153774],
       [-0.28001648,  1.03798438,  0.97368731, ..., -0.2532849 ,
        -1.2189321 , -0.65153774],
       [ 1.34010849,  1.560773  ,  1.58639956, ..., -0.2532849 ,
         0.82039024, -0.65153774],
       ...,
       [ 0.40166401, -0.18185576, -0.16420688, ..., -0.2532849 ,
        -1.2189321 , -0.65153774],
       [ 0.87611921, -0.87890726, -0.86444946, ..., -0.2532849 ,
         0.82039024, -0.65153774],
       [-1.15566543,  1.03798438,  0.97368731, ..., -0.2532849 ,
        -1.2189321 , -0.65153774]])

In [51]:
X_test_scaled

array([[-0.70842162, -1.31456446, -1.3896314 , ..., -0.2532849 ,
        -1.2189321 ,  1.53483051],
       [ 0.04861352,  0.16666999,  0.27344473, ..., -0.2532849 ,
        -1.2189321 ,  1.53483051],
       [ 0.09536131, -1.31456446, -1.30210108, ..., -0.2532849 ,
         0.82039024, -0.65153774],
       ...,
       [-0.47398494, -0.18185576, -0.16420688, ..., -0.2532849 ,
         0.82039024,  1.53483051],
       [-0.0951185 , -0.44325007, -0.33926753, ..., -0.2532849 ,
        -1.2189321 , -0.65153774],
       [-0.42165532, -0.2689872 , -0.33926753, ...,  3.94812332,
         0.82039024, -0.65153774]])

#### We separate the target variable from the features by dropping the 'Personal Loan' column from the features and storing it in a separate variable named 'y.' --> Next, we use scikit-learn's train_test_split function to split the dataset into 70% train and 30% test sets. We set the test_size parameter to 0.3 to indicate a 30% test set size, and we set the random_state parameter to 42 to ensure reproducibility of the results. --> Finally, we use scikit-learn's StandardScaler to standardize the columns of the train and test sets using z-score scaling. We fit the scaler on the train set and transform both the train and test sets using the fitted scaler. The standardized train and test sets are stored in variables named X_train_scaled and X_test_scaled, respectively.

# 9. Train and test a Logistic Regression model to predict the likelihood of a liability customer buying personal loans. Display the train and test accuracy scores.

In [53]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Create a Logistic Regression model object
model = LogisticRegression()

# Train the model on the standardized train set
model.fit(X_train_scaled, y_train)

# Predict the train and test sets using the trained model
y_train_pred = model.predict(X_train_scaled)
y_test_pred = model.predict(X_test_scaled)

# Compute the accuracy scores of the train and test sets
train_accuracy = accuracy_score(y_train, y_train_pred)
test_accuracy = accuracy_score(y_test, y_test_pred)

# Print the accuracy scores of the train and test sets
print('Train accuracy:', train_accuracy)
print('Test accuracy:', test_accuracy)

Train accuracy: 0.9522857142857143
Test accuracy: 0.9486666666666667


#### We then create a Logistic Regression model object named "model." --> Next, we train the model on the standardized train set using the model.fit() method. --> We then use the trained model to predict the train and test sets using the model.predict() method, and store the predicted values in variables named y_train_pred and y_test_pred, respectively. --> Finally, we compute the accuracy scores of the train and test sets using scikit-learn's accuracy_score() method, and store the scores in variables named train_accuracy and test_accuracy, respectively. We print the train and test accuracy scores using the print() function.

# 10. Print the confusion matrix and classification report for the model and write your conclusions on the results.

In [54]:
from sklearn.metrics import confusion_matrix, classification_report

# Compute the confusion matrix and classification report of the test set
confusion_mat = confusion_matrix(y_test, y_test_pred)
classification_rep = classification_report(y_test, y_test_pred)

# Print the confusion matrix and classification report of the test set
print('Confusion matrix:\n', confusion_mat)
print('Classification report:\n', classification_rep)

Confusion matrix:
 [[1323   20]
 [  57  100]]
Classification report:
               precision    recall  f1-score   support

           0       0.96      0.99      0.97      1343
           1       0.83      0.64      0.72       157

    accuracy                           0.95      1500
   macro avg       0.90      0.81      0.85      1500
weighted avg       0.95      0.95      0.95      1500



'''In this code, we first import the necessary libraries, i.e., scikit-learn's confusion_matrix and classification_report.
We then compute the confusion matrix and classification report of the test set using the confusion_matrix() and classification_report() methods, respectively. We pass in the actual test set labels (y_test) and the predicted test set labels (y_test_pred) as arguments to both methods.
Finally, we print the confusion matrix and classification report of the test set using the print() function. The confusion matrix shows the number of true positives, false positives, true negatives, and false negatives, while the classification report shows precision, recall, F1-score, and support for both classes (0 and 1).
Based on the results, we can see that the model has an overall accuracy of around 95% on both the train and test sets. The confusion matrix shows that the model correctly predicts a high number of true negatives and true positives, while misclassifying only a few instances as false positives or false negatives.
The classification report shows that the model has high precision, recall, and F1-score for both classes (0 and 1), indicating that it performs well in identifying both customers who are likely to buy personal loans and those who are not. Overall, the results suggest that the Logistic Regression model is a good fit for the data and can effectively predict the likelihood of a liability customer buying personal loans'''